File: bindings/soth-node/instrumentation/anthropic.js
// Anthropic Node SDK auto-instrumentation.
//
// Patches `Messages.create` on the typed messages module of
// `@anthropic-ai/sdk`. Anthropic's Node API is OpenAI-shaped except
// for the `system` parameter (separate field, not a system message)
// and tools (using `input_schema` rather than `parameters`).
//
//   client.messages.create({
//     model: 'claude-3-5-sonnet-latest',
//     messages: [{ role: 'user', content: '...' }],
//     system: '...',
//     stream: true|false,
//     tools: [{ name, description, input_schema }],
//     max_tokens: 1024,
//   })

const { wrapMethod, revertAll } = require('./_base.js');

const PROVIDER = 'anthropic';
const _patches = [];

function apply() {
  let anthropic;
  try {
    anthropic = require('@anthropic-ai/sdk');
  } catch (_) {
    return false;
  }

  // Anthropic's npm SDK has rotated typed-resource paths a few times
  // in its 0.x series. Try multiple candidates for version tolerance.
  const candidates = [
    () => require('@anthropic-ai/sdk/resources/messages'),
    () => require('@anthropic-ai/sdk/resources/messages/messages'),
  ];

  let messagesModule = null;
  for (const loader of candidates) {
    try {
      messagesModule = loader();
      break;
    } catch (_) {
      continue;
    }
  }
  if (!messagesModule) return false;

  const targets = [];
  const Messages = messagesModule.Messages;
  if (Messages && typeof Messages.prototype?.create === 'function') {
    targets.push([Messages, 'create']);
  }
  // Anthropic's beta module exports `MessagesBeta` with the same
  // `create` shape; patch when present.
  for (const exportName of Object.keys(messagesModule)) {
    const cls = messagesModule[exportName];
    if (
      typeof cls === 'function'
      && cls?.prototype
      && typeof cls.prototype.create === 'function'
      && !targets.some(([t]) => t === cls)
    ) {
      targets.push([cls, 'create']);
    }
  }

  for (const [target, methodName] of targets) {
    const patch = wrapMethod(target, methodName, {
      providerName: PROVIDER,
      buildCall,
      chunkExtractor,
    });
    if (patch) _patches.push(patch);
  }

  return _patches.length > 0;
}

function revert() {
  revertAll(_patches);
  _patches.length = 0;
}

function buildCall(args) {
  const opts = args?.[0] ?? {};
  const model = opts.model ?? '';
  const rawMessages = Array.isArray(opts.messages) ? opts.messages : [];

  const messages = rawMessages.map((m) => {
    const role = typeof m?.role === 'string' ? m.role : 'user';
    let content = m?.content ?? '';
    if (Array.isArray(content)) {
      // Multi-modal / tool-use content blocks: flatten text parts.
      content = content
        .map((p) => (p && typeof p === 'object' ? p.text ?? '' : ''))
        .filter(Boolean)
        .join(' ');
    } else if (content == null) {
      content = '';
    }
    return { role, content: String(content) };
  });

  const tools = [];
  if (Array.isArray(opts.tools)) {
    for (const t of opts.tools) {
      if (!t || typeof t !== 'object' || typeof t.name !== 'string') continue;
      tools.push({
        name: t.name,
        description: t.description ?? null,
        // Anthropic uses `input_schema` rather than OpenAI's `parameters`.
        parametersJson: stableStringify(t.input_schema ?? {}),
      });
    }
  }

  const call = {
    provider: PROVIDER,
    model: String(model),
    messages,
    stream: Boolean(opts.stream),
  };
  if (tools.length) call.tools = tools;
  if (opts.system) call.system = String(opts.system);
  return call;
}

function chunkExtractor(chunk) {
  // Anthropic streaming: events of type `content_block_delta`,
  // `message_delta`, `message_stop`. Text deltas live on
  // `chunk.delta.text`.
  try {
    const eventType = chunk?.type;
    if (eventType === 'content_block_delta') {
      const delta = chunk?.delta;
      const text = delta?.text ?? delta?.partial_json ?? null;
      return { deltaContent: text, finishReason: null };
    }
    if (eventType === 'message_delta') {
      const stop = chunk?.delta?.stop_reason ?? null;
      return { deltaContent: null, finishReason: stop };
    }
    if (eventType === 'message_stop') {
      return { deltaContent: null, finishReason: 'stop' };
    }
  } catch (_) { /* fall through */ }
  return { deltaContent: null, finishReason: null };
}

function stableStringify(obj) {
  if (obj === null || typeof obj !== 'object') return JSON.stringify(obj);
  if (Array.isArray(obj)) return `[${obj.map(stableStringify).join(',')}]`;
  const keys = Object.keys(obj).sort();
  const pairs = keys.map((k) => `${JSON.stringify(k)}:${stableStringify(obj[k])}`);
  return `{${pairs.join(',')}}`;
}

module.exports = {
  apply,
  revert,
  _buildCall: buildCall,
  _chunkExtractor: chunkExtractor,
};

File: bindings/soth-py/python/soth/instrumentation/_anthropic.py
"""Anthropic auto-instrumentation.

Patches `anthropic.resources.messages.Messages.create` and
`anthropic.resources.messages.AsyncMessages.create`.

Anthropic's request shape:

    client.messages.create(
        model="claude-...",
        messages=[{"role": "user", "content": "..."}],
        system="...",       # SEPARATE FIELD — not in messages
        stream=True/False,
        tools=[{"name": ..., "description": ..., "input_schema": {...}}],
        max_tokens=...,
    )

Differs from OpenAI in:
- `system` is a separate parameter (not a message with role=system)
- Tools use `{name, description, input_schema}` shape (vs OpenAI's nested)
"""

from __future__ import annotations

import logging
from typing import Any, Optional

from ._base import Patch, revert_all, wrap_method

logger = logging.getLogger("soth.instrumentation.anthropic")

PROVIDER = "anthropic"
_patches: list[Patch] = []


def apply() -> bool:
    """Patch Anthropic's messages classes. Returns True on success,
    False if the anthropic package isn't importable."""
    try:
        from anthropic.resources import messages as msg_module
    except ImportError:
        return False

    targets: list[tuple[type, str]] = []

    sync_cls = getattr(msg_module, "Messages", None)
    if sync_cls is not None:
        targets.append((sync_cls, "create"))

    async_cls = getattr(msg_module, "AsyncMessages", None)
    if async_cls is not None:
        targets.append((async_cls, "create"))

    for target, method_name in targets:
        patch = wrap_method(
            target,
            method_name,
            provider_name=PROVIDER,
            build_call=_build_call,
            chunk_extractor=_chunk_extractor,
        )
        if patch is not None:
            _patches.append(patch)

    return bool(_patches)


def revert() -> None:
    revert_all(_patches)
    _patches.clear()


def _build_call(args: tuple, kwargs: dict) -> dict[str, Any]:
    """Extract an LlmCall dict from Anthropic's `create(...)` args."""
    model = kwargs.get("model")
    raw_messages = kwargs.get("messages") or []
    messages: list[dict[str, str]] = []
    for m in raw_messages:
        role = m.get("role", "user") if isinstance(m, dict) else "user"
        content = m.get("content", "") if isinstance(m, dict) else ""
        # Anthropic content may be a string OR a list of ContentBlock
        # objects (`{"type": "text", "text": "..."}`). Flatten.
        if isinstance(content, list):
            parts = []
            for p in content:
                if isinstance(p, dict):
                    text = p.get("text", "")
                    if text:
                        parts.append(text)
            content = " ".join(parts)
        elif content is None:
            content = ""
        messages.append({"role": str(role), "content": str(content)})

    tools_kwarg = kwargs.get("tools") or []
    tools: list[dict[str, Any]] = []
    for t in tools_kwarg:
        if not isinstance(t, dict):
            continue
        name = t.get("name")
        if not name:
            continue
        tools.append(
            {
                "name": str(name),
                "description": t.get("description"),
                "parameters_json": _stable_json_dumps(t.get("input_schema", {})),
            }
        )

    call: dict[str, Any] = {
        "provider": PROVIDER,
        "model": str(model) if model else "",
        "messages": messages,
        "stream": bool(kwargs.get("stream", False)),
    }
    if "system" in kwargs and kwargs["system"]:
        call["system"] = kwargs["system"]
    if tools:
        call["tools"] = tools
    return call


def _chunk_extractor(chunk: Any) -> tuple[Optional[str], Optional[str]]:
    """Anthropic streaming yields a sequence of MessageStreamEvent
    objects (`MessageStartEvent`, `ContentBlockDeltaEvent`,
    `MessageStopEvent`, ...). The text deltas live on
    `ContentBlockDeltaEvent.delta.text`.

    Returns `(text, finish_reason)` where `finish_reason` is set on
    the terminal `MessageDeltaEvent` when present.
    """
    try:
        event_type = getattr(chunk, "type", None)
        if event_type == "content_block_delta":
            delta = getattr(chunk, "delta", None)
            if delta is not None:
                text = getattr(delta, "text", None) or getattr(delta, "partial_json", None)
                return (text, None) if text else (None, None)
        if event_type == "message_delta":
            delta = getattr(chunk, "delta", None)
            stop = getattr(delta, "stop_reason", None) if delta else None
            return (None, stop)
        if event_type == "message_stop":
            return (None, "stop")
    except (AttributeError, TypeError):
        pass
    return (None, None)


def _stable_json_dumps(obj: Any) -> str:
    import json

    try:
        return json.dumps(obj, sort_keys=True, separators=(",", ":"))
    except (TypeError, ValueError):
        return repr(obj)
